File: app-android/src/main/java/org/mtransit/android/ad/OnInitializationCompleteListener.java
package org.mtransit.android.ad;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.ads.initialization.AdapterStatus;
import com.google.android.gms.ads.initialization.InitializationStatus;

import org.mtransit.android.ad.banner.BannerAdManager;
import org.mtransit.android.commons.MTLog;
import org.mtransit.android.ui.view.common.IActivity;

import java.lang.ref.WeakReference;
import java.util.Map;

class OnInitializationCompleteListener implements com.google.android.gms.ads.initialization.OnInitializationCompleteListener, MTLog.Loggable {

	private static final String LOG_TAG = AdManager.class.getSimpleName() + ">" + OnInitializationCompleteListener.class.getSimpleName();

	@NonNull
	@Override
	public String getLogTag() {
		return LOG_TAG;
	}

	@NonNull
	private final BannerAdManager bannerAdManager;
	@NonNull
	private final WeakReference<IActivity> activityWR;

	OnInitializationCompleteListener(@NonNull BannerAdManager bannerAdManager, @Nullable IActivity activity) {
		this.bannerAdManager = bannerAdManager;
		this.activityWR = new WeakReference<>(activity);
	}

	@Override
	public void onInitializationComplete(@NonNull InitializationStatus initializationStatus) {
		MTLog.d(this, "onInitializationComplete()");
		final IActivity activity = this.activityWR.get();
		if (activity != null) {
			this.bannerAdManager.setupBannerAd(activity, false);
		}
		Map<String, AdapterStatus> statusMap = initializationStatus.getAdapterStatusMap();
		for (String adapterClass : statusMap.keySet()) {
			AdapterStatus status = statusMap.get(adapterClass);
			MTLog.d(this, "Adapter name: %s, Description: %s, Latency: %d",
					adapterClass,
					status == null ? null : status.getDescription(),
					status == null ? null : status.getLatency()
			);
		}
	}
}

File: app-android/src/main/java/org/mtransit/android/ad/OnInitializationCompleteListener.kt
package org.mtransit.android.ad

import com.google.android.gms.ads.initialization.InitializationStatus
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener
import org.mtransit.android.ad.banner.BannerAdManager
import org.mtransit.android.commons.MTLog
import org.mtransit.android.commons.MTLog.Loggable
import org.mtransit.android.ui.view.common.IActivity
import java.lang.ref.WeakReference

internal class OnInitializationCompleteListener(
    private val bannerAdManager: BannerAdManager,
    private val activityWR: WeakReference<IActivity>,
    private val adScreenFragmentWR: WeakReference<IAdScreenFragment>,
) : OnInitializationCompleteListener, Loggable {

    constructor(
        bannerAdManager: BannerAdManager,
        activity: IActivity?,
        adScreenFragment: IAdScreenFragment?,
    ) : this(
        bannerAdManager,
        WeakReference<IActivity>(activity),
        WeakReference<IAdScreenFragment>(adScreenFragment),
    )

    companion object {
        private val LOG_TAG = AdManager::class.java.getSimpleName() + ">" + OnInitializationCompleteListener::class.java.getSimpleName()
    }

    override fun getLogTag() = LOG_TAG

    override fun onInitializationComplete(initializationStatus: InitializationStatus) {
        val activity = this.activityWR.get()
        if (activity != null) {
            this.bannerAdManager.refreshBannerAdStatus(activity, this.adScreenFragmentWR.get(), false)
        }
        val statusMap = initializationStatus.adapterStatusMap
        for (adapterClass in statusMap.keys) {
            val status = statusMap[adapterClass]
            MTLog.d(this, "Adapter name: $adapterClass, Description: ${status?.description}, Latency: ${status?.latency}")
        }
    }
}
